File: remove_node_from_end_of_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        front = back = head

        while n > 0:
            front = front.next
            n -= 1

        if not front:
            return head.next

        while front.next:
            front = front.next
            back = back.next

        back.next = back.next.next
        return head

File: removeNodeFromEndOfLinkedList.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
  /**
   * @param {ListNode} head
   * @param {number} n
   * @return {ListNode}
   */
  removeNthFromEnd(head, n) {
    let front = head;
    let back = head;

    while (n > 0) {
      front = front.next;
      n--;
    }

    if (front === null) {
      return head.next;
    }

    while (front.next !== null) {
      front = front.next;
      back = back.next;
    }

    back.next = back.next.next;
    return head;
  }
}
